tensorflow/compiler/xla/service/collective_ops_utils.cc

```diff
@@ -53,12 +53,24 @@ std::optional<ReductionKind> MatchReductionInstruction(
 std::optional<ReductionKind> MatchReductionComputation(
     const HloComputation* computation) {
   namespace m = match;
-  const HloInstruction* root = computation->root_instruction();
+  HloInstruction* root = computation->
```

Comment:
Not required anymore?

Reply:
Overall, I wonder whether we can add a separate pass to transform add(convert(x), convert(y)) = convert(add(x,y)) and then we don't need to special case match this in this pass. This can also help other passes like reduce-scatter reassociation. That is, instead of matching these converts, we can push these add "down" through the converts so that the existing pass can then find more matching cases.

Reply:
Is this safe for up-converts?

Reply:
I think having a pass for the add(convert(x), convert(y)) case is more general, can we also merge allreduce and reducescatter reassociations into 1 pass? they seem to share the same logic.

Reply:
I am checking if there is an existing pass we can use. @vinodgro you are right that we what we are doing here is downgrading the fp32 add (after the converts) to a bf16 add, so its not numerically correct. May be we should have a flag for this?

Reply:
The bf16 pass should add converts to the inputs of add later on, but that doesn't seem to be very robust. If we merge allreduce and reducescatter passes, then we can embed this logic in the same pass instead of passing the flag.

Reply:
I discussed this with a few XLA/GPU folks and the consensus is that this is particularly unsafe and we should not allow this in XLA. Can you check why the add of the 2 all-reduces in the example you are trying to optimize is fp32. Can the model be adjusted to make that addition in bf16 as well?

Reply:
I believe the add is for weight update of sharded embedding tables which has to be done in fp32 for stability. Changing user code is not always viable. I could move the pattern matcher to an utility function so allreduce and reducescatter can share?

Reply:
> be done in fp32 for stability.

But the concern is that this transform will convert that add to bf16, which is not correct. So the consensus is that we should not do this numerically unsafe transformation in XLA.

Reply:
i meant the pattern matcher in this pr, this will transform:
```
             allreduce
                      \
                     (cast)
                            \
allreduce- (cast) -           add
```
into

```
       (cast)
              \
(cast) -      add - (cast) - allreduce
```

Reply:
In the example you have for the test case, there is 4 bf16 all-reduce, which are then converted to fp32 and added. Can you clarify what your transform will do? I suppose to preserve the numerical accuracy, it has to cast all the inputs of the ARs to fp32,  and then do an fp32 all-reduce. So it needs to promote the bf16 all-reduce to fp32 as a part of the reassociation. Do you agree? And in that case is is still beneficial?

Reply:
If the input types of allreduce and add are different, it will inject a cast between add and allreduce, so the example will output:
4 inputs -> cast to f32 -> add->add->add -> cast to bf16 -> allreduce -> cast to f32.
I guess the doubt is whether it's safe to perform allreduce on summed results.
Promoting allreduce to f32 is roughly the same as halving number of allreduces, the hope is this will detect more patterns recursively to reduce AR count by >2x in order to show perf benefit, i.e in our testing, it could effectively reduce 3 grad ARs to 1